### PR TITLE
build: update master to 4.0.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

**Comment by sdelamo** [comment]
`compileJava` fails with: 


```
Unable to find method ''io.micronaut.inject.ast.Element io.micronaut.inject.ast.ClassElement.annotate(java.lang.Class)''
'io.micronaut.inject.ast.Element io.micronaut.inject.ast.ClassElement.annotate(java.lang.Class)'

Gradle's dependency cache may be corrupt (this sometimes occurs after a network connection timeout.)

Re-download dependencies and sync project (requires network)
The state of a Gradle build process (daemon) may be corrupt. Stopping all Gradle daemons may solve this problem.

Stop Gradle build processes (requires restart)
Your project may be using a third-party plugin which is not compatible with the other plugins in the project or the version of Gradle requested by the project.

In the case of corrupt Gradle processes, you can also try closing the IDE and then killing all Java processes.

```

**Comment by graemerocher** [comment]
@sdelamo this last error is because the 4.0.x core build is broken and the latest published snapshot isn't the latest

File: security-annotations/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+plugins {
+    id("io.micronaut.build.internal.security-module")
+}
+dependencies {
+    implementation("io.micronaut:micronaut-core-processor")
```
Comment (graemerocher) [inline security-annotations/build.gradle.kts:5]: should  be `compileOnly` because this module is transitively used by the runtime modules

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: security-annotations/build.gradle.kts (hunk 1 above)
Comment (sdelamo) [reply security-annotations/build.gradle.kts]: changed to `compileOnly`

**Comment by graemerocher** [comment]
With the latest snapshot it fails with:

```
> Task :security:compileJava
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
Note: Creating bean classes for 43 type elements
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
Note: Creating bean classes for 1 type elements
/Users/graemerocher/dev/micronaut/security/security/src/main/java/io/micronaut/security/authentication/Authentication.java:18: error: package com.fasterxml.jackson.databind.annotation does not exist
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
                                                ^
/Users/graemerocher/dev/micronaut/security/security/src/main/java/io/micronaut/security/authentication/Authentication.java:33: error: cannot find symbol
@JsonDeserialize(as = ClientAuthentication.class)
 ^
  symbol: class JsonDeserialize
/Users/graemerocher/dev/micronaut/security/security/src/main/java/io/micronaut/security/endpoints/introspection/IntrospectionController.java:29: error: package io.micronaut.json does not exist
import io.micronaut.json.JsonMapper;
```

This is because Jackson is no longer a transitive

**Comment by timyates** [comment]
I think some of the failures here are because micronaut-servlet isn't working with the current 4.0.0-SNAPSHOT again

File: settings.gradle [hunk 2]
```diff
@@ -24,8 +24,8 @@ include "test-suite"
 include "test-suite-geb"
 include "test-suite-keycloak"
 include "test-suite-keycloak-16"
-include "test-suite-tomcat"
-include "test-suite-tomcat-groovy"
+//include "test-suite-tomcat"
```
Comment (sdelamo) [inline settings.gradle:27]: I have temporary comment out this test modules until servlet is ready for 4.0.0-SNAPSHOT

File: config/accepted-api-changes.json [hunk 3]
```diff
@@ -43,5 +43,135 @@
     "type": "io.micronaut.security.token.jwt.validator.JwtClaimsValidator",
     "member": "Field PREFIX",
     "reason": "Removed deprecated constant for Micronaut 4"
+  },
+  {
+    "type": "io.micronaut.security.rules.$InterceptUrlMapRule$Definition$Reference",
+    "member": "Implemented interface io.micronaut.inject.BeanType",
+    "reason": "Provide a human readable reason for the change"
```
Comment (timyates) [inline config/accepted-api-changes.json:50]:
```suggestion
    "reason": "Changed in Micronaut 4"
```

**Review by timyates — APPROVED** [review]
👍 

Maybe removing moshi and json-unit could be a different commit

**Comment by timyates** [comment]
Some of the smells may be valid...  THere's one deprecation that I think should be removed

File: gradle/libs.versions.toml [hunk 4]
```diff
@@ -18,10 +18,13 @@ velocity-engine-core = "2.3"
 bouncycastle = "1.70"
 javax-activation = "1.2.0"
 kotlin = "1.7.20"
+jsonunit = "2.36.0"
+moshi = "1.14.0"
```
[the remark this answers is not in the document]
Comment (timyates) [reply gradle/libs.versions.toml:22]: Yes, as Groovy doesn't care about map order when comparing maps

[the remark this answers is not in the document]
Comment (sdelamo) [reply gradle/libs.versions.toml]: resolved via https://github.com/micronaut-projects/micronaut-security/pull/1127/commits/8c114342ca05b708ff90ad74708a5e23fd2ec811

File: gradle.properties [hunk 5]
```diff
@@ -21,6 +21,8 @@ testssecurityjava=security/src/test/java/io/micronaut/docs
 testssecurity=security/src/test/groovy/io/micronaut/docs
 testssecurityjwt=security-jwt/src/test/groovy/io/micronaut/docs
 testssecuritysession=security-session/src/test/groovy/io/micronaut/docs
+testsuitegeb=test-suite-geb/src/test/groovy/io/micronaut/
 
 org.gradle.caching=true
 org.gradle.jvmargs=-Xmx1g
+systemProp.predictiveTestSelection=false
```
[the remark this answers is not in the document]
Comment (sdelamo) [reply gradle.properties]: resolved via https://github.com/micronaut-projects/micronaut-security/pull/1127/commits/dfbfd4f5cc10bb376f19cbb62774b9f0ee5ba983

**Review by timyates — APPROVED** [review]

**Pull request merged**